Support for AT86RF215 radio on OpenMoteB board

9morano opened the pull request:
## AT86RF215
The PR adds support for the [AT86RF215](https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-42415-WIRELESS-AT86RF215_Datasheet.pdf) radio available on OpenMoteB devices. 
Support for the radio was previously requested in #1102 .
The radio contains 2 RF cores (2.4 GHz and sub GHz) and supports the following PHY modes: MR-FSK, MR-O-QPSK, and MR-OFDM . 

## Driver
The driver was designed to provide support for only one core at the time (only 2.4 GHz was tested).
The driver currently implements the legacy O-QPSK, but if there is a need, I will be happy to try to help implement other modulations.
The driver is mainly designed to support TSCH MAC mode, but CSMA works as well.

### Usage

`#define OPENMOTEB_CONF_USE_ATMEL_RADIO     1`

9morano committed:
Drivers for Atmel's AT86RF215 radio

9morano committed:
Added support for AT86RF215 radio on OpenMoteB board

9morano committed:
Example application for selecting  between radios on the OpenMoteB board

pjonsson commented on arch/platform/openmote/dev/antenna.c:
```diff
@@ -73,6 +79,20 @@ antenna_internal(void)
 }
 /*---------------------------------------------------------------------------*/
+void
+antenna_select_atmel(void)
```
What is the difference between these two functions?

9morano replied:
Both OpenMote-CC2538 and OpenMote-B boards contain an antenna switch.

- On OpenMote-CC2538, the switch is used to connect the output of the radio to either the internal (chip antenna) or external antenna (SMA connector).

- For OpenMote-B, on the other hand, the switch selects which output of the radio goes to the SMA connector - either Atmel's AT86RF215 radio or TI CC2538 radio.

On both boards, the RF switch is connected to the same GPIO pins - I have created new functions (which basically do the same thing as the original ones) just to avoid any misunderstandings about function names.

At this point I am not sure if I should create a separate file for each board, or if we should use the originally defined functions and just add some documentation describing the differences between the boards?

pjonsson replied:
Initially I thought they set/clear the same pins, but they set/clear the opposite pins, so ignore the comment.

9morano replied:
I will update (rebase) this PR when the #2684 is accepted

pjonsson commented on arch/platform/openmote/openmote-b/Makefile.openmote-b:
```diff
@@ -0,0 +1,2 @@
+CONTIKI_TARGET_SOURCEFILES += at86rf215.c at86rf215-arch.c
```
Isn't `at86rf215.c` part of the module added in the next line?

pjonsson commented on examples/platform-specific/openmote-b/hello-world/Makefile:
```diff
@@ -0,0 +1,8 @@
+MAKE_MAC = MAKE_MAC_TSCH
+MODULES += os/services/shell
+
+CONTIKI = ../../../..
```
No need to define CONTIKI in Makefiles any longer.

9morano replied:
I didn't know that - just took an example from other examples, all of which contain the definition of CONTIKI.

9morano committed:
Removed unnecessary sourcefile

pjonsson commented on arch/dev/radio/at86rf215/at86rf215.h:
```diff
@@ -0,0 +1,128 @@
+
+/* Basic radio configuration */
+typedef struct {
+    uint16_t address;
```
Typedef something for the type of address, otherwise things will be hard to update when the address gets more bits in the future.

pjonsson commented on arch/dev/radio/at86rf215/at86rf215.h:
```diff
@@ -0,0 +1,128 @@
+    //uint16_t crc;
+    uint8_t lqi;
+    int8_t rssi;
+	rtimer_clock_t timestamp;
```
Unless something in the code mandates the order of the members, put timestamp first to avoid padding between struct elements.

pjonsson commented on arch/dev/radio/at86rf215/at86rf215.h:
```diff
@@ -0,0 +1,128 @@
+/** The NETSTACK data structure for the AT86RF215 driver */
+extern const struct radio_driver at86rf215_driver;
+
+#endif /* AT86RF215_H_ */
```
I think there's a missing newline at the end of several files, look for the red ring and fix those.

pjonsson commented on arch/platform/openmote/dev/at86rf215-arch.c:
```diff
@@ -0,0 +1,169 @@
+    SPIX_BUF(AT86RF215_SPI_INSTANCE) = b;
+    SPIX_WAITFOREOTx(AT86RF215_SPI_INSTANCE);
+    SPIX_WAITFOREORx(AT86RF215_SPI_INSTANCE);
+    b = SPIX_BUF(AT86RF215_SPI_INSTANCE);
```
No need to assign to b, just `return SPIX_BUF(..);` directly instead.

pjonsson commented on arch/platform/openmote/dev/at86rf215-arch.c:
```diff
@@ -0,0 +1,169 @@
+    /* Turn the radio off */
+    GPIO_CLR_PIN(AT86RF215_PWR_PORT_BASE , AT86RF215_PWR_PIN_MASK );
+    GPIO_CLR_PIN(AT86RF215_RSTN_PORT_BASE ,AT86RF215_RSTN_PIN_MASK );
+    for(uint16_t delay=0; delay<0xA2C2; delay++);
```
The compiler might optimize these away, there should be some delay function to call if that is what is required.

9morano committed:
AT86RF215: minor improvements

9morano committed:
OpenMoteB: AT86RF215 radio improvements

arurke commented:
Thanks @9morano! A few comments:
* Add building with the necessary defines in `tests/02-compile-arm-ports/Makefile` to get some test coverage?
* It seems to me the example is quite redundant as it does not showcase anything new besides enabling the radio. Assuming there is nothing special with this radio that needs an example I would suggest you remove the example and update [the docs](https://docs.contiki-ng.org/en/develop/doc/platforms/openmote-cc2538.html) to reflect your changes and explain how to enable the radio.
* Remove unrelated changes (is the changes in platform.c in related? changes of UART and SPI in board.h?)
* You seem to make the sub-ghz radio the default now - I am not familiar with the platform/board but I would advise against changing the default behavior without a thorough rationale

9morano committed:
Removed redundant example application

9morano committed:
Removed unrelated changes

9morano committed:
doc: update description of OpenMote platforms

9morano committed:
Added compilation test for OpenMote-B platform with AT86RF215 radio

9morano commented:
@arurke thank you for the comments!

- I have added a compilation test for OpenMote-B platform with AT86RF215 transceiver.
- I agree with the comment. I have removed the redundant example application and updated the documentation on OpenMote platforms (both OpenMote-CC2538 and OpenMote-B boards).
- Unrelated changes in the board.h were removed (changes to the UART). I would leave the changes to the SPI defines, as they emphasize the difference between SPI0 and SPI1. Regarding the platfrom.c, the `board_init()` function was moved into `platform_init_stage_two()` so that the radio's signal path (RF switch) is configured before the radio is initialized.
- The OpenMote-B board provides two radio chips: Texas Instruments' CC2538 and Atmel's AT86RF215. By default, the board operates with the CC2538 radio using the 2.4 GHz RF interface. If you set `OPENMOTEB_CONF_USE_ATMEL_RADIO` to 1, the board will use the AT86RF215 radio with new drivers, otherwise the board will continue using the old implementation. Although the AT86RF215 radio supports both 2.4 GHz and sub-GHz RF interfaces, the current driver only supports the 2.4 GHz interface.

arurke commented on doc/platforms/openmote.md:
```diff
@@ -26,28 +20,42 @@ It has the following key features:
 
-To start using Contiki-NG with the OpenMote-CC2538, the following is required:
+The port is organized as follows:
+* the derivers for CC2538 CPU are located in the `arch/cpu/cc2538` folder,
```
typo - derivers

arurke commented on doc/platforms/openmote.md:
```diff
@@ -26,28 +20,42 @@ It has the following key features:
+  * `dev` contains drivers for various sensors,
+  * `openmote-cc2538` includes configuration for the OpenMote-CC2538 board,
+  * `openmote-b` includes configuration for the OpenMote-B board.
+* the driver for AT86RF215 radio (available only on OpenMote B) is located in `arch/dev/radio/at86rf215`.
```
nitpicking - OpenMote-B ?

arurke commented on doc/platforms/openmote.md:
```diff
@@ -56,34 +64,42 @@ Instructions to flash for different OS are given below.
 * On Linux:
-    * Nodes can be programmed with TI's [UniFlash] tool. With UniFlash, use the file with `.elf` extension.
-    * Nodes can also be programmed via the serial boot loader in the cc2538. No extra software needs to be installed.
+    * Nodes can be programmed with TI's UniFlash tool. With UniFlash, use the file with `.elf` extension.
```
Is there an .elf extension output from the build? Is it not a .openmote or something like that?

9morano committed:
OpenMoteB: use CC2538 radio by  default

9morano committed:
Updated doc for OpenMote platform

9morano commented on arch/platform/openmote/openmote-b/board.h:
```diff
@@ -181,15 +182,62 @@
+ * @{
+ */
+#ifndef OPENMOTEB_CONF_USE_ATMEL_RADIO
+#define OPENMOTEB_USE_ATMEL_RADIO               1
```
I overlooked that, thanks!

joakimeriksson commented:
We intend to merge this soon - can you do a rebase? @9morano

joakimeriksson commented:
One final thing before we merge. Could you @9morano remove the CMSIS commit (the last one) as we already have a newer version now in develop?

joakimeriksson commented:
Thanks for quick responses @9morano!
